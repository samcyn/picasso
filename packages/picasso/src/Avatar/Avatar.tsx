import React, { FunctionComponent, HTMLAttributes } from 'react'
import cx from 'classnames'
import { makeStyles, Theme } from '@material-ui/core/styles'
import {
  StandardProps,
  SizeType,
  OmitInternalProps,
  JssProps
} from '@toptal/picasso-shared'

import Image from '../Image'
import { Props as ImageProps } from '../Image/Image'
import Logo from '../Logo'
import Typography from '../Typography'
import getNameInitials from '../utils/get-name-initials'
import styles from './styles'
import { AVATAR_INITIALS_LIMIT } from '../utils/constants'
import { Profile16 } from '..'

type VariantType = 'square' | 'portrait' | 'landscape'

export interface Props extends StandardProps, HTMLAttributes<HTMLDivElement> {
  /** Alt text */
  alt?: string
  /** User full name to display initials on the avatar */
  name?: string
  /**
   * Size
   * @default xsmall
   */
  size?: SizeType<'xxsmall' | 'xsmall' | 'small' | 'medium' | 'large'>
  /** Photo url */
  src?: string
  /**
   * Variant of the avatar shape
   * @default square
   */
  variant?: VariantType
}

const useStyles = makeStyles<Theme, Props>(styles, { name: 'PicassoAvatar' })

const isBrowserSupportsObjectFit = 'objectFit' in document.documentElement.style

const renderLogo = ({
  classes,
  src,
  size,
  name
}: Pick<Props, 'src' | 'size' | 'name'> & JssProps) => {
  if (!src || !size || !name || ['small', 'xsmall', 'xxsmall'].includes(size)) {
    return null
  }

  return (
    <div className={classes.logoContainer}>
      <Logo emblem variant='white' className={classes.logo} />
    </div>
  )
}

const renderInitials = ({
  classes,
  src,
  name
}: Pick<Props, 'src' | 'name'> & JssProps) => {
  if (src || !name) {
    return null
  }

  const initials = getNameInitials(name)

  return (
    <Typography
      className={cx(classes.text, classes.absoluteCenter, {
        [classes.textCapLimit]: initials.length >= AVATAR_INITIALS_LIMIT
      })}
      invert
    >
      {initials}
    </Typography>
  )
}

const renderIcon = ({
  classes,
  name,
  size = 'xsmall',
  src
}: Pick<Props, 'src' | 'name' | 'size'> & JssProps) => {
  if (src || name) {
    return null
  }

  return (
    <Profile16
      className={cx(classes.absoluteCenter, classes[`${size}Icon`])}
      color='white'
    />
  )
}

// You will be surprised, but it's a IE11 fix for `object-fit: cover` for images
const IE11Image = ({ style, src, ...rest }: OmitInternalProps<ImageProps>) => (
  <div
    style={{
      backgroundImage: `url(${src})`,
      backgroundSize: 'cover',
      backgroundPosition: 'center center',
      ...style
    }}
    {...rest}
  />
)

export const Avatar: FunctionComponent<Props> = props => {
  const {
    alt,
    src,
    className,
    name,
    size = 'xsmall',
    style,
    variant = 'square',
    ...rest
  } = props

  const classes = useStyles(props)

  const sizeClassName = classes[size]
  const variantClassName = classes[variant]

  const InputComponent = isBrowserSupportsObjectFit ? Image : IE11Image

  return (
    <div {...rest} className={cx(classes.root, sizeClassName)}>
      {src && (alt || name) ? (
        <InputComponent
          alt={alt || String(name)}
          className={cx(
            classes.image,
            variantClassName,
            sizeClassName,
            classes.clippedCorner,
            className
          )}
          src={src}
          style={style}
        />
      ) : (
        <div
          className={cx(
            classes.textContainer,
            variantClassName,
            sizeClassName,
            classes.clippedCorner,
            className
          )}
        />
      )}
      {renderInitials({ classes, src, name })}
      {renderIcon({ classes, src, name, size })}
      {renderLogo({ classes, src, size, name })}
    </div>
  )
}

Avatar.defaultProps = {
  size: 'xsmall',
  variant: 'square'
}

Avatar.displayName = 'Avatar'

export default Avatar
