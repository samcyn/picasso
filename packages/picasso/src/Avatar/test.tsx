import React from 'react'
import { render } from '@toptal/picasso/test-utils'
import { OmitInternalProps } from '@toptal/picasso-shared'

import Avatar, { Props } from './Avatar'

const renderAvatar = (props: OmitInternalProps<Props>) => {
  const { alt, name, src } = props

  return render(<Avatar alt={alt} name={name} src={src} />)
}

describe('Avatar', () => {
  it('renders with initials', () => {
    const { container } = renderAvatar({ name: 'Jacqueline Roque' })

    expect(container).toMatchSnapshot()
  })

  it('renders with a long name', () => {
    const { container } = renderAvatar({
      name: 'Jacqueline Roque Bailey Armstrong'
    })

    expect(container).toMatchSnapshot()
  })

  it('renders with an image', () => {
    const { container } = renderAvatar({
      alt: 'Photo alt text',
      src: 'http://example.png',
      name: 'Jacqueline Roque'
    })

    expect(container).toMatchSnapshot()
  })

  it('renders with a placeholder icon', () => {
    const { container } = renderAvatar({})

    expect(container).toMatchSnapshot()
  })
})
