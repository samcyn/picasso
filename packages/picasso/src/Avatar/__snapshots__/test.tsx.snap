// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders with a long name 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="PicassoAvatar-root PicassoAvatar-xsmall"
    >
      <div
        class="PicassoAvatar-textContainer PicassoAvatar-square PicassoAvatar-xsmall PicassoAvatar-clippedCorner"
      />
      <p
        class="MuiTypography-root PicassoTypography-bodyInherit PicassoTypography-invert PicassoAvatar-text PicassoAvatar-centeredContent PicassoAvatar-textCapLimit MuiTypography-body1"
      >
        JRB
      </p>
    </div>
  </div>
</div>
`;

exports[`Avatar renders with a placeholder icon 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="PicassoAvatar-root PicassoAvatar-xsmall"
    >
      <div
        class="PicassoAvatar-textContainer PicassoAvatar-square PicassoAvatar-xsmall PicassoAvatar-clippedCorner"
      />
      <svg
        class="PicassoSvgProfile16-root PicassoAvatar-centeredContent PicassoAvatar-xsmallIcon PicassoSvgProfile16-white"
        style="min-width: 16px; min-height: 16px;"
        viewBox="0 0 16 16"
      >
        <path
          d="M5.639 9.408a5 5 0 114.723 0A7.003 7.003 0 0115 16h-1a6 6 0 10-12 0H1a7.003 7.003 0 014.639-6.592zM8 9a4 4 0 100-8 4 4 0 000 8z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Avatar renders with an image 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="PicassoAvatar-root PicassoAvatar-xsmall"
    >
      <img
        alt="Photo alt text"
        class="PicassoImage-root PicassoAvatar-image PicassoAvatar-square PicassoAvatar-xsmall PicassoAvatar-clippedCorner"
        src="http://example.png"
      />
    </div>
  </div>
</div>
`;

exports[`Avatar renders with initials 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="PicassoAvatar-root PicassoAvatar-xsmall"
    >
      <div
        class="PicassoAvatar-textContainer PicassoAvatar-square PicassoAvatar-xsmall PicassoAvatar-clippedCorner"
      />
      <p
        class="MuiTypography-root PicassoTypography-bodyInherit PicassoTypography-invert PicassoAvatar-text PicassoAvatar-centeredContent MuiTypography-body1"
      >
        JR
      </p>
    </div>
  </div>
</div>
`;
